### PR TITLE
Add possibility to encode the state param as UrlEncodedBase64

### PR DESCRIPTION
**Pull request by brezinajn**
## Description

This is continuation of #2239

Added a feature to optionally Base64Url encode the state param.
The feature is behind a configuration flag, so the default behavior remains unchanged.

## Motivation and Context

We're facing an issue where the state parameter gets truncated by the provider, leading to an incorrect application state after login. It's probably [this](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/adfs.go#L59) issue but in our case with an oidc provider.

## How Has This Been Tested?

We used a build from this commit as a drop-in replacement for 7.5.0 on our testing environment. Works as expected.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.
- [x] I have written tests for my code changes.

### CODE REVIEW

**Review by tuunit — APPROVED** [review]

**Review by JoelSpeed — DISMISSED** [review]
I'm ok with this. I think though when we go through the alpha config migration to GA, I would probably suggest changing up the bool to an enum.

Something like `stateEncoding: Plain | Base64` would be more flexible and allow us to add additional options in the future should we choose to do so.

**Comment by JoelSpeed** [comment]
Looks like there's a conflict that will need to be resolved before we merge

**Comment by brezinajn** [comment]
> Looks like there's a conflict that will need to be resolved before we merge

Resolved

**Review by JoelSpeed — DISMISSED** [review]

**Comment by JoelSpeed** [comment]
Can we fix the linter issue please?

**Comment by tuunit** [comment]
Hi @brezinajn,

Hope you are doing well. We would like to get this merged. The following linting issue is holding us back:
```bash
Error: oauthproxy.go:1205:9: indent-error-flow: if block ends with a return statement, so drop this else and outdent its block (revive)
	} else {
		return rawString
	}
```

File: oauthproxy.go [hunk 1]
```diff
@@ -1194,18 +1198,29 @@ func checkAllowedEmails(req *http.Request, s *sessionsapi.SessionState) bool {
...
 // original application redirect
-func encodeState(nonce string, redirect string) string {
-	return fmt.Sprintf("%v:%v", nonce, redirect)
+func encodeState(nonce string, redirect string, encode bool) string {
+	rawString := fmt.Sprintf("%v:%v", nonce, redirect)
+	if encode {
+		return base64.RawURLEncoding.EncodeToString([]byte(rawString))
+	} else {
+		return rawString
+	}
```
Comment (tuunit) [inline oauthproxy.go:1207]:
```suggestion
	}
	return rawString
```

**Review by tuunit — CHANGES_REQUESTED** [review]

**Comment by brezinajn** [comment]
> Hi @brezinajn,
> 
> Hope you are doing well. We would like to get this merged. The following linting issue is holding us back:
> 
> ```shell
> Error: oauthproxy.go:1205:9: indent-error-flow: if block ends with a return statement, so drop this else and outdent its block (revive)
> 	} else {
> 		return rawString
> 	}
> ```

Hi, sorry I'm on my vacation, so I'm not at my PC currently. Committed your suggestion. Hope it resolves the issue. Should there be another issue, please let me know. Thanks

**Comment by oshmyrko** [comment]
Hi guys! Any updates on this?

**Review by kvanzuijlen — APPROVED** [review]

**Comment by kvanzuijlen** [comment]
> Hi guys! Any updates on this?

This should be picked up during the next round of merging hopefully.

**Review by JoelSpeed — APPROVED** [review]

**Pull request merged**